redhat-marketplace/redhat-marketplace-operator
# Feature/file server final merge copy

## Overview
This pr creates two controllers:
- `meterdefinition_install_controller` : 
    - watches for new subscriptions, looks up the InstalledCSV, and installs the appropriate meterdefinitions from the file server
- `deployment_config_controller` : 
    - Deploys:
        - DeploymentConfig: [DeploymentConfig docs](https://docs.openshift.com/container-platform/4.8/applications/deployments/what-deployments-are.html), which has two containers:
            - kube-proxy: [kube-proxy repo](https://github.com/brancz/kube-rbac-proxy).
            - https file server built in golang: [meterdefintion catalog server](https://github.com/redhat-marketplace/metering-extensions/tree/>main/rhm-file-server).
        - ImageStream: [ImageStream docs](https://docs.openshift.com/container-platform/4.8/openshift_images/image-streams-manage.html)
        - Service for the DeploymentConfig
    - The `DeploymentConfig` watches for new images to the file server image repo and rolls out a new deployment if the image in the repo changes
    -  the deploymentconfig controller watches the `DeploymentConfig` status for new deployments. If a new deployment is detected it will perform a `sync` which will call the file server deployed by the deploymentconfig and compare the meterdefinitions in the latest file server image with the metetrdefinitions on the cluster.
    - A `sync` handles CRUD operations for the latest meterdefs coming from the file server
        - Create: if a new meterdef is in the latest image it gets added to the cluster.
        - Update: if a meterdefinition from the latest image contains an update to it's Annotations or Spec we find the correlating meterdefinition on the cluster and apply the update.
        - Delete: if a meterdefinition gets removed from the catalog and that meterdefinition is still on the cluster.

### Meterdefinitions Applied from the Meterdefintion File Server Repository:
The deployment config controller will apply `Community Meterdefinitions` and `System Meterdefinitions`.
  - `Community Meterdefinitions`: are created by ISVs and should be added to the [meterdefinition catalog](https://github.com/redhat-marketplace/metering-extensions/tree/sub-annotations/meterdefinition-catalog). After their Meterdefinition is added, they should submit a pr with their addition to the meterdefinition catalog
  - `System Meterdefinitions`: are created by the RHM team and get applied with every operator that gets installed on customer clusters
    - the file server uses the [product-index.json](https://github.com/redhat-marketplace/metering-extensions/blob/sub-annotations/admin/product-index.json) hosted in the file server repository as a single source of truth to determine whether system meterdefinitions should get applied for a csv on cluster. 
    - the `product-index.json` will be maintained by admins and is not meant to be updated by community meterdefinition developers

### The file server image:
- the image is currently set to `quay.io/mxpaspa/rhm-meterdefinition-file-server:sub-annotations-1.0.0` in `config.go`
- this is the image built from the `sub-annotations` branch on the file server: https://github.com/redhat-marketplace/metering-extensions/tree/sub-annotations

## How to test:
#### Use operators from RHM or your own catalog:
reference your subscription and find the `name` (package name) and `source` (catalog source) fields:
```YAML
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  labels:
    "marketplace.redhat.com/operator": "true"
name: joget-dx-operator-rhmp
namespace: openshift-redhat-marketplace
spec:
  channel: alpha
  installPlanApproval: Automatic
  name: joget-dx-operator-rhmp
  source: redhat-marketplace
  sourceNamespace: openshift-marketplace
  startingCSV: joget-dx-openshift-operator.v0.0.14
```

- update the product-index.json file: [product-index.json](https://github.com/redhat-marketplace/metering-extensions/blob/sub-annotations/admin/product-index.json) with the package name and catalog source of the operator. This will allow the file server to identify your operator on the cluster and apply system meterdefinitions

```JSON
{
    "indexes":[
        {
            "subscription.name" : "joget-dx-operator-rhmp",
            "subscription.source" : "redhat-marketplace"
        }
    ] 
}
```

- update your community meterdefinitions in the meterdefinition catalog with the annotations shown below. This will allow the file server to identify your operator on the cluster and apply community meterdefinitions.

```YAML
apiVersion: marketplace.redhat.com/v1beta1
kind: MeterDefinition
metadata:
  name: joget-dx-operator-1
  namespace: 
  annotations:
    "versionRange": "<=0.5.55"
    "subscription.name": "joget-dx-operator-rhmp"
    "subscription.source" : "redhat-marketplace"
```
`versionRange`: 
- This tells the file server what versions of your csv a particular meterdefinition supports. In the example above the `joget-dx-operator-1` meterdefinition can meter all version less than or equal to version 0.5.55 of the joget csv.
- the file server uses versioning provided by [sever](https://github.com/Masterminds/semver#basic-comparisons)

After adding your package name and catalog source, you can submit a pr with your changes. Once the pr is merged a build process will start that will update the image repository for the file server.

For local testing see [Test Locally](#test-locally). You can add your meterdefinitions or update `product-index.json` then push to your own image repository
#### Use the provided test catalog
The file server repository already has an updated `product-index.json` and community meterdefinitions with correct annotations for the catalog source and subscription provided. 
_catalog source_:
```YAML
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: max-test-catalog
  namespace: openshift-redhat-marketplace
spec:
  sourceType: grpc
  image: quay.io/mxpaspa/memcached-ansible-index:1.0.1 
```

_subscription_:
```YAML
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  labels:
    "marketplace.redhat.com/operator": "true"
  name: memcached-operator-rhmp
  namespace: openshift-redhat-marketplace
spec:
  channel: alpha
  installPlanApproval: Manual
  name: memcached-operator-rhmp
  source: max-test-catalog
  sourceNamespace: openshift-redhat-marketplace
```

### Test Locally:
- clone: https://github.com/redhat-marketplace/metering-extensions/tree/sub-annotations
- from the project root run `docker build -t {your-repo:tag} .` and push to your personal repository
- in `config.go` change: 
```
DeploymentConfig  string `env:"RELATED_IMAGE_DEPLOYMENT_CONFIG" envDefault:"{your-repo:tag}"`
```
- run `make skaffold-dev`
- check that the image stream is pulling from your repository by running `oc describe imagestream`. You should see something like:

```Name:			rhm-meterdefinition-file-server
Namespace:		openshift-redhat-marketplace
Created:		9 hours ago
Labels:			app=rhm-meterdefinition-file-server
Annotations:		openshift.io/image.dockerRepositoryCheck=2021-10-05T00:38:49Z
Image Repository:	{your-repo}
Image Lookup:		local=false
Unique Images:		1
Tags:			1

v1
  updates automatically from registry quay.io/mxpaspa/rhm-meterdefinition-file-server:sub-annotations-1.0.0
    will use insecure HTTPS or HTTP connections

  * quay.io/mxpaspa/rhm-meterdefinition-file-server@sha256:5db19bfcfcd1df209523ebeb4a1f1e6c8ec871d98e97916a012ee6e0866270e8
      2 hours ago
```
